overlord/devicemgmtstate/devicemgmtmgr.go

```diff
@@ -0,0 +1,561 @@
+	// up to this point.
+	PendingAckToken string `json:"pending-ack-token"`
+
+	// ReadyResponses are response-message assertions ready to send in the next poll.
+	// Cleared after successful transmission.
+	ReadyResponses map[string]store.Message `json:"ready-responses"`
+
+	// Timestamp of last poll
+	LastPolled time.Time `json:"last-polled"`
+}
+
+// DeviceMgmtManager handles device management operations.
+type DeviceMgmtManager struct {
+	state    *state.State
+	signer   ResponseMessageSigner
+	handlers map[string]MessageHandler
+}
+
+// Manager creates a new DeviceMgmtManager.
+func Manager(state *state.State, runner *state.TaskRunner, signer ResponseMessageSigner) (*DeviceMgmtManager, error) {
+	m := &DeviceMgmtManager{
+		state:    state,
+		signer:   signer,
+		handlers: make(map[string]MessageHandler),
+	}
+
+	m.handlers["confdb"] = &ConfdbMessageHandler{}
+
+	runner.AddHandler("poll-messages", m.doPollMessages, nil)
+	runner.AddHandler("dispatch-messages", m.doDispatchMessages, nil)
+	runner.AddHandler("validate-message", m.doValidateMessage, nil)
+	runner.AddHandler("apply-message", m.doApplyMessage, nil)
+	runner.AddHandler("await-subsystem-change", m.doAwaitSubsystemChange, nil)
+	runner.AddHandler("queue-response", m.doQueueResponse, nil)
+
+	return m, nil
+}
+
+// getState retrieves the current device management state, initializing if not present.
+func (m *DeviceMgmtManager) getState() (*DeviceMgmtState, error) {
+	var ms DeviceMgmtState
+	err := m.state.Get(deviceMgmtStateKey, &ms)
+	if err != nil {
+		if errors.Is(err, state.ErrNoState) {
+			return &DeviceMgmtState{
+				PendingMessages: make(map[string]*PendingMessage),
+				Sequences:       make(map[string]*Sequence),
+				ReadyResponses:  make(map[string]store.Message),
+			}, nil
+		}
+
+		return nil, err
+	}
+
+	return &ms, nil
+}
+
+// setState persists the device management state.
+func (m *DeviceMgmtManager) setState(ms *DeviceMgmtState) {
+	m.state.Set(deviceMgmtStateKey, ms)
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *DeviceMgmtManager) Ensure() error { // TODO: add manager-level feature flag?
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	ms, err := m.getState()
+	if err != nil {
+		return err
+	}
+
+	for _, chg := range m.state.Changes() {
+		if chg.Kind() == deviceMgmtCycleChangeKind && !chg.Status().Ready() {
+			return nil
+		}
+	}
+
+	ts := state.NewTaskSet()
+	var dispatch *state.Task
+	for _, msg := range ms.PendingMessages {
+		if msg.ChangeID == "" {
+			dispatch = m.state.NewTask("dispatch-messages", "Dispatch message(s) to subsystems")
+			ts.AddTask(dispatch)
+			break
+		}
+	}
+
+	if m.canPoll(ms.LastPolled) {
+		poll := m.state.NewTask("poll-messages", "Poll store for device management messages")
+		ts.WaitFor(poll)
+		ts.AddTask(poll)
+	}
+
+	if len(ts.Tasks()) > 0 {
+		chg := m.state.NewChange(deviceMgmtCycleChangeKind, "Process device management messages")
+		chg.AddAll(ts)
+		chg.Set("dispatch-queued", dispatch != nil)
+	}
```

Comment:
So my understanding is that we're creating the dispatch first if there are pending messages, then we create the polling and set the dispatch to wait on it, then we set a flag on the change if the dispatch exists. Then after doing the polling we create a dispatch if there are messages and we didn't create one before. The construction is a bit confusing compared to the actual order of operations. 
It might be simpler to create the polling task and then only check for pending messages once, in the polling task. Either that or we schedule the dispatch after the polling anyway and if there are no messages, it's a no-op. Regardless I think it would simplify things quite a bit